### PR TITLE
Task 2

### PR DESCRIPTION
**Pull request by gorr2012**
1. Task-2
2. Deadline/Done: 25.10.2021 / 25.10.2021
3. Score: 5/5

## Evaluation criteria Task 2
- [x] 2.Task 2.1 is partially implemented (w/o getAutoSuggestUsers or other methods).
- [x] 3.Task 2.1 is fulfilled to the full extent.
- [x] 4.Task 2.1eslint rules are applied.
- [x] 5.Task 2.2is fulfilled to the full extent; validation package is used.
- [x] 5*.Consider to use Typescript.

### CODE REVIEW

File: src/services/usersServices.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  const userIdToUpdate = memory.findIndex((user) => user.id === userToSave.id);
+  const userToUpdete = memory[userIdToUpdate];
+  memory[memory.findIndex((user) => user.id === userToSave.id)] = {
+    ...userToUpdete,
+    ...userToSave,
+  };
+};
+
+export const findAndDelete = async (id: string) => {
+  const userId = memory.findIndex((user) => user.id === id);
```
Comment (aliskar) [inline src/services/usersServices.ts:20]: Could you double-check what will happen if you try to get access to none existing user? 🙄

File: src/services/memory.ts [hunk 2]
```diff
@@ -0,0 +1,3 @@
+import { User } from '../types/types';
+
+export const memory: User[] = [];
```
Comment (aliskar) [inline src/services/memory.ts]: It's better to create not empty storage with some mocked data.

File: src/routes/usersRoutes.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+import { createUser, getAllUsers, getUserById, updateUser, deleteUser, handleOtherPaths } from '../controllers';
+
+export const usersRoutes = Router();
+const validator = createValidator();
+
+usersRoutes
+  .get('/', getAllUsers)
+  .get('/:userId', getUserById)
+  .post('/', json(), validator.body(userSchema), createUser)
+  .put('/', json(), validator.body(userSchema), updateUser)
```
Comment (aliskar) [inline src/routes/usersRoutes.ts]:
```suggestion
  .post('/', validator.body(userSchema), createUser)
  .put('/', validator.body(userSchema), updateUser)
```
You could add one body parser at the top-level of your app (app.ts).
```
app.use(json());
```

File: src/services/usersServices.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+  memory[userId].isDeleted = true;
+};
+
+export const getAutoSuggestUsers = async (loginSubstring: string, limit?: number) => {
+  const usersAllNotDeleted = await getAllUsersNotDeleted();
+  const filteredUsers = usersAllNotDeleted
+    .sort((a, b) => a.login.localeCompare(b.login))
+    .filter((user) => user.login.includes(loginSubstring));
```
Comment (aliskar) [inline src/services/usersServices.ts]: What do you think about case sensitivity when we try to suggest users? 🤔

File: src/validate/validate.ts [hunk 5]
```diff
@@ -0,0 +1,10 @@
+import Joi from 'joi';
+
+export const userSchema = Joi.object({
+  login: Joi.string().required(),
+  password: Joi.string()
+    .regex(/\d/, '"password should contains at least one number"')
+    .regex(/[A-Za-z]/, '"assword should contains at least one letter"')
```
Comment (aliskar) [inline src/validate/validate.ts]:
```suggestion
    .regex(/[A-Za-z]/, '"password should contains at least one letter"')
```

File: src/controllers/updateUser.ts [hunk 6]
```diff
@@ -0,0 +1,10 @@
+import { update } from '../services/usersServices';
+import { Response, Request } from 'express';
+
+const updateUser = async (req: Request, res: Response) => {
```
Comment (aliskar) [inline src/controllers/updateUser.ts:4]: Doesn't work. 🙄

File: src/routes/usersRoutes.ts [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+  .delete('/:userId', deleteUser)
+  .get('*', handleOtherPaths);
```
Comment (aliskar) [inline src/routes/usersRoutes.ts]: In which case this route will work? Seems to me, it's redundant. All `get` requests with one `/` activate this one `get('/:userId', getUserById)`.

**Review by aliskar — CHANGES_REQUESTED** [review]
Need to fix user update method.

File: src/routes/usersRoutes.ts (hunk 3 above)
Comment (gorr2012) [reply src/routes/usersRoutes.ts]: done

File: src/validate/validate.ts (hunk 5 above)
Comment (gorr2012) [reply src/validate/validate.ts]: done

File: src/services/memory.ts (hunk 2 above)
Comment (gorr2012) [reply src/services/memory.ts]: done

File: src/services/usersServices.ts (hunk 4 above)
Comment (gorr2012) [reply src/services/usersServices.ts]: done

File: src/routes/usersRoutes.ts (hunk 7 above)
Comment (gorr2012) [reply src/routes/usersRoutes.ts]: done

**Review by aliskar — APPROVED** [review]

File: src/controllers/updateUser.ts (hunk 6 above)
Comment (gorr2012) [reply src/controllers/updateUser.ts:4]: Now it works )

**Review by aliskar — APPROVED** [review]
well done

**Pull request merged**